dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/jsonpatch/JsonPatchManager.java

```diff
@@ -0,0 +1,96 @@
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.jsonpatch;
+
+import org.hisp.dhis.commons.jackson.jsonpatch.JsonPatch;
+import org.hisp.dhis.commons.jackson.jsonpatch.JsonPatchException;
+import org.hisp.dhis.schema.Property;
+import org.hisp.dhis.schema.Schema;
+import org.hisp.dhis.schema.SchemaService;
+import org.hisp.dhis.system.util.ReflectionUtils;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+
+/**
+ * @author Morten Olav Hansen
+ */
+@Service
+public class JsonPatchManager
+{
+    private final ObjectMapper jsonMapper;
+
+    private final SchemaService schemaService;
+
+    public JsonPatchManager(
+        ObjectMapper jsonMapper,
+        SchemaService schemaService )
+    {
+        this.jsonMapper = jsonMapper;
+        this.schemaService = schemaService;
+    }
+
+    @Transactional
+    @SuppressWarnings( "unchecked" )
+    public <T> T apply( JsonPatch patch, T object )
+        throws JsonPatchException
+    {
+        if ( patch == null || object == null )
+        {
+            return null;
+        }
+
+        Schema schema = schemaService.getSchema( object.getClass() );
+
+        JsonNode node = jsonMapper.valueToTree( object );
+
+        for ( Property property : schema.getProperties() )
+        {
+            if ( property.isCollection() )
+            {
+                Object data = ReflectionUtils.invokeMethod( object, property.getGetterMethod() );
+                ((ObjectNode) node).set( property.getCollectionName(), jsonMapper.valueToTree( data ) );
+            }
+        }
```

Comment:
Could be extracted to a method. 
Also I wonder why this is needed. Shouldn't the `valueToTree` already work for collections as well?
Maybe a comment should explain why this is needed?

Reply:
I will add a comment to it, yes I was hoping it would work just with the old version also (but it doesn't).